face_detection.py

```diff
@@ -0,0 +1,190 @@
+import torch
+import numpy as np
+import os
+from PIL import Image, ImageDraw
+import cv2
+from facenet_pytorch.models.mtcnn import fixed_image_standardization
+from matplotlib import pyplot as plt
+from facenet_pytorch import MTCNN, InceptionResnetV1
+import torchvision
+from torchvision import transforms
+from tqdm import tqdm
+from tkinter import Tk, filedialog
+import subprocess
+import platform
+from datetime import datetime
+from pathlib import Path
+import pandas as pd
+from smile_net import Net
+
+
+class FaceObject:
+    # center is numpy array, to present the last center point [x,y]
+    def __init__(self, box: list, pred: int, idx: int, v_len: int):
+        self.current_center = np.array([(box[0] + box[2])/2, (box[1] + box[3])/2])
+        self.predictions_in_frames = [None] * v_len
+        self.predictions_in_frames[idx] = pred
+        self.coordinates_in_frames = [None] * v_len
+        self.coordinates_in_frames[idx] = box
+
+
+def assign(face_objects_list: list, boxes, predictions, idx, v_len) -> list:
+    max_assign_distance = 50
+    if len(boxes) != len(predictions):
+        print("error, faces and predictions should have same length")
+        return face_objects_list
+
+    for box, prediction in zip(boxes, predictions):
+        center = np.array([(box[0] + box[2])/2, (box[1] + box[3])/2])
+
+        min_distance = max_assign_distance
+        closest_face = None
+        for face in face_objects_list:
+            distance = np.linalg.norm(face.current_center - center)
+            if distance < min_distance:
+                min_distance = distance
+                closest_face = face
+
+        if closest_face:
+            closest_face.current_center = center
+            closest_face.predictions_in_frames[idx] = prediction
+            closest_face.coordinates_in_frames[idx] = box
+        else:
+            face = FaceObject(box, prediction, idx, v_len)
+            face_objects_list.append(face)
+
+    return face_objects_list
+
+
+def correct(face_objects_list: list):
+    min_frame_appearance = 10
+    window_size = 10
+    sensitivity = 0
+    for face in face_objects_list:
+        count_frame_appearance = sum(x is not None for x in face.predictions_in_frames)
+        if count_frame_appearance < min_frame_appearance:
+            face_objects_list.remove(face)
+            continue
+
+        for i, x in enumerate(face.predictions_in_frames):
+            stri = str(x.item()) if x is not None else ''
+            print('frame #', i, 'frame detect:', stri)
+            # print(x.item(), i)
+
+        corrected_predictions: list = face.predictions_in_frames
+        half_window_size = int(window_size/2)
+        for i in range(half_window_size, len(face.predictions_in_frames) - half_window_size):
+            if sum(x is None for x in
```

Comment:
`sum(x is None for x in` seems like a duplication - can it be a small function with the end of the array as input?